### PR TITLE
Lazy load components as they are resolved, rather than on injection

### PR DESCRIPTION
**Pull request by timriley**
Over in #33 we discussed how it might be better to load a system's components whenever they are resolved, rather than just when they were injected, since it'll allow the lazy loading behaviour to work in more circumstances.

Here's a first take at it. I initially tried to do it with a custom "resolver" for the `Dry::Container` that `Dry::System::Container` inherits from, but it didn't quite give me what I need, since Dry::Container's resolver class doesn't actually have access to the container object itself, just its internal storage hash, and I needed access to the system object so I could call `#load_component`.

Anyway, I realised that all I _really_ needed was a custom `#resolve` on `Dry::System::Container` with a single extra line added, to call `#load_component`.

What do you think, @solnic, @AMHOL? This seem fair to incorporate?

The only thing I don't like is the begin/rescue. I felt like it would be better for a failure to load the component to be manifested as an actual `Dry::Container::Error` "nothing registered" error rather than a dry-system `ComponentLoadError`, since the auto-loading is really meant to be a side-effect of calling `#resolve`, not the actual outcome. Can you think of a way to do this more nicely?

### CODE REVIEW

**Comment by timriley** [comment]
One other benefit of this approach: we can totally rip out our custom `Injector` class, which is nice!

**Comment by blelump** [comment]
Hi @timriley ,

I've landed here, because it seems I'm having similar problems, so I've pulled this PR to see how it works. It indeed lazy loads dependencies, however what if a container imports another container (actually this is what I'm doing)? Without finalization, import doesn't provide any dependency and yet, it doesn't know that 'parent' container should load that particular dependency, so in the end it cannot find suitable one and blows up.

**Comment by AMHOL** [comment]
I'm not sure about the error suppression TBH, if we are implicitly loading I think that I'd rather see a loading error rather than a missing key error.

**Comment by solnic** [comment]
Lazy-loading should only be used when container is not finalized. I think eventually we need a way of configuring containers with different loading strategies so that in case of a finalized container (as in, application's runtime environment), we simply expect all objects to be registered already.

**Comment by timriley** [comment]
I updated this since @solnic's last comment. What do you think now? Happy to see it merged?

File: lib/dry/system/container.rb [hunk 1]
```diff
@@ -413,6 +413,13 @@ def root
           config.root
         end
 
+        # @api public
+        def resolve(key)
+          load_component(key) unless frozen?
+
+          super(key)
```
Comment (solnic) [inline lib/dry/system/container.rb]: just `super` :)

**Comment by solnic** [comment]
This is a nice code removal. I only left one minor comment. Let's get this merged and see how it works for us.

**Pull request merged**